flink-core/src/main/java/org/apache/flink/statistics/OperatorStatistics.java

```diff
@@ -0,0 +1,154 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.statistics;
+
+import com.clearspring.analytics.stream.cardinality.CardinalityMergeException;
+import com.clearspring.analytics.stream.cardinality.HyperLogLog;
+import com.clearspring.analytics.stream.cardinality.ICardinality;
+import com.clearspring.analytics.stream.cardinality.LinearCounting;
+import org.apache.flink.statistics.heavyhitters.IHeavyHitter;
+import org.apache.flink.statistics.heavyhitters.LossyCounting;
+import org.apache.flink.statistics.heavyhitters.CountMinHeavyHitter;
+import org.apache.flink.statistics.heavyhitters.HeavyHitterMergeException;
+
+import java.io.Serializable;
+import java.util.Map;
+
+/**
+ * Data structure that encapsulates statistical information of data that has only been processed by one pass
+ * This statistical information is meant to help determine the distribution of the data that has been processed
+ * in an operator so that we can determine if it is necessary to repartition the data
+ *
+ * The statistics to be gathered are configurable and represented by a {@link OperatorStatisticsConfig} object.
+ *
+ * The information encapsulated in this class is min, max, a structure enabling estimation of count distinct and a
+ * structure holding the heavy hitters along with their frequency.
+ *
+ */
+public class OperatorStatistics implements Serializable {
+
+	OperatorStatisticsConfig config;
+
+	Object min;
+	Object max;
+	ICardinality countDistinct;
+	IHeavyHitter heavyHitter;
+	long cardinality = 0;
+
+	public OperatorStatistics(OperatorStatisticsConfig config) {
+		this.config = config;
+		if (config.countDistinctAlgorithm.equals(OperatorStatisticsConfig.CountDistinctAlgorithm.LINEAR_COUNTING)) {
+			countDistinct = new LinearCounting(OperatorStatisticsConfig.COUNTD_BITMAP_SIZE);
+		}
+		if(config.countDistinctAlgorithm.equals(OperatorStatisticsConfig.CountDistinctAlgorithm.HYPERLOGLOG)){
+			countDistinct = new HyperLogLog(OperatorStatisticsConfig.COUNTD_LOG2M);
+		}
+		if (config.heavyHitterAlgorithm.equals(OperatorStatisticsConfig.HeavyHitterAlgorithm.LOSSY_COUNTING)){
+			heavyHitter =
+					new LossyCounting(OperatorStatisticsConfig.HEAVY_HITTER_FRACTION, OperatorStatisticsConfig.HEAVY_HITTER_ERROR);
+		}
+		if (config.heavyHitterAlgorithm.equals(OperatorStatisticsConfig.HeavyHitterAlgorithm.COUNT_MIN_SKETCH)){
+			heavyHitter =
+					new CountMinHeavyHitter(OperatorStatisticsConfig.HEAVY_HITTER_FRACTION,
+											OperatorStatisticsConfig.HEAVY_HITTER_ERROR,
+											OperatorStatisticsConfig.HEAVY_HITTER_CONFIDENCE,
+											OperatorStatisticsConfig.HEAVY_HITTER_SEED);
+		}
+	}
+
+	public void process(Object tupleObject){
```

Comment:
It looks like this method expected to be called for each passing element if statistics collection is enabled. This could add significant processing overhead. Would it make sense to add an optional skip interval n that processes only every n-th element (except from increasing the `cardinality` counter.

Also it makes sense to perform as many checks outside of the function as possible. The type of the data elements is known before execution. So it can be checked if a data type allows statistics collection at program composition or optimization time.